### ISSUE TITLE
CI Request: Automatically add version upper-bounds to reverse dependencies as necessary when a package updates

### ISSUE THREAD

**Issue by gridbugs** [ocurrent/opam-repo-ci#217]
When a package is updated with an API-breaking change, reverse dependencies which specify their dependency on the package without an upper bound on its version number will break. A reasonable way to address these breakages would be to set the upper bound in the version constraints in these packages to the previously released version of the newly-updated package. I think it would be useful to have the PR suggest these changes automatically.

**Comment by gridbugs** [comment]
@mseri in response to https://github.com/ocaml/opam-repository/issues/23694#issuecomment-1518106351, what's the current process for fixing version upper bounds on revdeps? Is it more involved than just setting them to the previous version of the package being updated?

**Comment by mseri** [comment]
Yes, sometimes we detect implicit dependencies, for examples or implicit conflicts due to some other packages or conflicts with depopts. 

There is also the lower bounds check which may issues (especially on packages that were not tested before its introduction)

**Comment by shonfeder** [comment]
@punchagan and I have discussed support for this kind of workflow as an extension to our ongoing work of factoring out the revdeps tests into a tool that can be run locally. The idea would be to support something akin to a `--promote` functionality that would respond to failing revdeps in with the addition of upper bound limits.

More recently, following a discussion with @pitag-ha, we've discussed support for semi-automating the process of submitting PRs to fix downstream packages, which would be an even nicer way to address breaking API changes (when feasible). This would amount to support for semi-automated "(sub)ecosystem migrations".

An lighter lift would be functionality that compliments the automated addition of upper-bound range with pings to maintainers alerting them of the breaking API change (via automated issue creation via supported forges, perhaps with a fallback to email).

**Comment by shubhamkumar13** [comment]
@shonfeder I'm interested in picking up this issue.

**Comment by shonfeder** [comment]
Awesome! Let me know if/when you want to talk over design ideas or constraints! I'd also be happy to read any notes you might share on how you're thinking about the problems :)

From our chats, I have a suggestion about the smallest addition her that would get us most of the benefit:

When revdeps fail, there are three possible reasons:

1. The something is broken with the revdep itself (e.g., its source has gone missing).
2. There is an unintended breaking API change in the package being published. 
3. There is an intended breaking API change in the package being published. This is the case that is the focus of this issue but really, we need to consider all 3 together to solve this sensibly.

Each has a separate remedy:

1. The problematic revdep package should be fixed. (But, sometimes this is not at all easy to do, and it would also help to be able to track that the package is known to be broken, so we could avoid a bunch of noise by having a way of marking known broken packages and filter the revdeps to test accordingly.)
2. In this case, the package being published needs to be fixed, and the current errors in revdeps should be suitable to indicate this.
3. In this case, we already have a semi-automated repair system via [`opam admin add-constraint`](https://opam.ocaml.org/doc/man/opam-admin.html), but people don't know about it!

IMO, the smallest fix here that would bring the most benefit is just having clear documentation about how to handle revdep failures, and then making it easy for package authors to discover this documentation (e.g., via https://github.com/ocurrent/opam-repo-ci/issues/373#issuecomment-2899067701). 

Currently, the only documentation we have on dealing with revdeps is https://github.com/ocaml/opam-repository/wiki/Revdeps .

Moreover, writing this documentation might be a good way to help you think thru all the aspects of the problem, and [documentation is (the best way to start) automation](https://cacm.acm.org/practice/documentation-is-automation/) :)

Once we have clear documentation that is easily discoverable by package authors, we could then see where the remaining pain points are, and focus more expensive and involved automation on fixing those.

WDYT?

**Comment by shubhamkumar13** [comment]
I agree with the conclusion that updating the documentation is a good first step.

**Comment by shubhamkumar13** [comment]
I was investigating *Reason #2* for unintended API breakage.

I found an existing solution in [Python](https://github.com/StardustDL/aexpy)

Which I am interested to look at. Lmk if this is in line with what you are referring to in #2.

**Comment by shonfeder** [comment]
Thanks, @shubhamkumar13 !

I'm not sure I understand how https://github.com/StardustDL/aexpy is relevant here. What would it give us that OCaml's type system already doesn't w/r/t breaking API detection?

AFAIU, we don't have a problem of understanding *whether or not* there is a breaking change to a package's API (relative to some revdep). We have a way to detect this: the revdep won't build or its tests will fail.

The question is, given the revdeps failure, what is causing it. Reasons (1), (2), and (3) are the three possible causes. AFAIK, there is no way to automate the decision between (2) and (3), since this is a question of the author's intent.

LMK if having a sync discussion on this would help!

**Comment by shubhamkumar13** [comment]
I think you are right aexpy is just detecting the breaking API change. 

The only thing one can do is let the author know.

I was wrong to assume this could be automated without a bot but I think the opan bot suggestion is the way to go forward.
